miner/worker_builder.go

```diff
@@ -0,0 +1,406 @@
+			return signalToErr(commitInterruptBundleCommit)
+
+		case core.ErrNonceTooLow:
+			// New head notification data race between the transaction pool and miner, shift
+			log.Error("Transaction with low nonce in the bundle", "sender", from, "nonce", tx.Nonce())
+			return signalToErr(commitInterruptBundleCommit)
+
+		case core.ErrNonceTooHigh:
+			// Reorg notification data race between the transaction pool and miner, skip account =
+			log.Error("Account with high nonce in the bundle", "sender", from, "nonce", tx.Nonce())
+			return signalToErr(commitInterruptBundleCommit)
+
+		case nil:
+			// Everything ok, collect the logs and shift in the next transaction from the same account
+			coalescedLogs = append(coalescedLogs, logs...)
+			env.tcount++
+			continue
+
+		default:
+			// Strange error, discard the transaction and get the next in line (note, the
+			// nonce-too-high clause will prevent us from executing in vain).
+			log.Error("Transaction failed in the bundle", "hash", tx.Hash(), "err", err)
+			return signalToErr(commitInterruptBundleCommit)
+		}
+	}
+
+	if !w.isRunning() && len(coalescedLogs) > 0 {
+		// We don't push the pendingLogsEvent while we are mining. The reason is that
+		// when we are mining, the worker will regenerate a mining block every second.
+		// In order to avoid pushing the repeated pendingLog, we disable the pending log pushing.
+
+		// make a copy, the state caches the logs and these logs get "upgraded" from pending to mined
+		// logs by filling in the block hash when the block was mined by the local miner. This can
+		// cause a race condition if a log was "upgraded" before the PendingLogsEvent is processed.
+		cpy := make([]*types.Log, len(coalescedLogs))
+		for i, l := range coalescedLogs {
+			cpy[i] = new(types.Log)
+			*cpy[i] = *l
+		}
+		w.pendingLogsFeed.Send(cpy)
+	}
+	return nil
+}
+
+// generateOrderedBundles generates ordered txs from the given bundles.
+// 1. sort bundles according to computed gas price when received.
+// 2. simulate bundles based on the same state, resort.
+// 3. merge resorted simulateBundles based on the iterative state.
+func (w *worker) generateOrderedBundles(
+	env *environment,
+	bundles []*types.Bundle,
+) (types.Transactions, *types.SimulatedBundle, error) {
+	// sort bundles according to gas price computed when received
+	sort.SliceStable(bundles, func(i, j int) bool {
+		priceI, priceJ := bundles[i].Price, bundles[j].Price
+
+		return priceI.Cmp(priceJ) >= 0
+	})
+
+	// recompute bundle gas price based on the same state and current env
+	simulatedBundles, err := w.simulateBundles(env, bundles)
+	if err != nil {
+		log.Error("fail to simulate bundles base on the same state", "err", err)
+		return nil, nil, err
+	}
+
+	// sort bundles according to fresh gas price
+	sort.SliceStable(simulatedBundles, func(i, j int) bool {
+		priceI, priceJ := simulatedBundles[i].BundleGasPrice, simulatedBundles[j].BundleGasPrice
+
+		return priceI.Cmp(priceJ) >= 0
```

Comment:
Use `>=` in stable sort will make it unstable. Should use `>` instead.

Check this example: https://go.dev/play/p/wwyLf916mrD.

Reply:
BTW, according to the doc https://pkg.go.dev/sort#SliceStable, it's recommend to use https://pkg.go.dev/slices#SortStableFunc instead.

Reply:
Refer doc, refine it, please review again.